src/handlers/miner_dkg_penalty_handler.erl

```diff
@@ -0,0 +1,232 @@
+
+
+-record(state, {bbas :: #{pos_integer() => hbbft_bba:bba_data()},
+                privkey :: tpke_privkey:privkey(),
+                members :: [ libp2p_crypto:pubkey_bin()],
+                f :: pos_integer(),
+                dkg_results :: [boolean(),...],
+                bba_results = #{} :: #{pos_integer() => 0 | 1},
+                done_called = false :: boolean(),
+                conf_sent = false :: boolean(),
+                delay :: non_neg_integer(),
+                height :: pos_integer(),
+                artifact,
+                signatures=[] :: [{libp2p_crypto:address(), binary()}]
+               }).
+
+init([Members, PrivKey, Signatures, Delay, Height]=_Args) ->
+    N = length(Members),
+    F = floor((N - 1) / 3),
+    BBAs = [ {I, hbbft_bba:init(PrivKey, N, F)} || I <- lists:seq(1, length(Members))],
+    DKGResults = lists:map(fun(Member) ->
+                                   lists:keymember(Member, 1, Signatures)
+                           end, Members),
+    {ok, #state{bbas=maps:from_list(BBAs),
+                privkey=PrivKey,
+                f=F,
+                delay=Delay,
+                height=Height,
+                members=Members,
+                dkg_results=DKGResults}}.
+
+handle_command(start, State) ->
+    {NewBBAs, OutMsgs} = lists:foldl(fun({BBAIndex, SawSig}, {BBAs, Msgs}) ->
+                                             Input = case SawSig of
+                                                         true -> 1;
+                                                         false -> 0
+                                                     end,
+                                             {NewBBA, BBAMsgs} = hbbft_bba:input(maps:get(BBAIndex, State#state.bbas), Input),
+                                             {maps:put(BBAIndex, NewBBA, BBAs), fixup_bba_msgs(BBAMsgs, BBAIndex)++Msgs}
+                                     end, {State#state.bbas, []}, lists:zip(lists:seq(1, length(State#state.members)), State#state.dkg_results)),
+    {reply, ok, OutMsgs, State#state{bbas=NewBBAs}};
+handle_command(stop, State) ->
+    lager:info("stop called without timeout"),
+    {reply, ok, [{stop, 0}], State}.
+
+handle_message(BinMsg, Index, State) when is_binary(BinMsg) ->
+    try binary_to_term(BinMsg) of
+        Msg ->
+            handle_message(Msg, Index, State)
+    catch _:_ ->
+            lager:warning("got truncated message: ~p:", [BinMsg]),
+            ignore
+    end;
+handle_message({conf, Signatures}, _Index, State)
+  when State#state.artifact /= undefined andalso State#state.done_called == false ->
+    GoodSignatures =
+        lists:foldl(
+          fun({Address, Signature}, Acc) ->
+                  %% only check signatures from members we have not already verified and have not already appeared in this list
+                  case {lists:keymember(Address, 1, Acc) == false andalso
```

Comment:
Maybe it doesn't matter but the order of operation here is flipped I think, membership check in state.members should happen first imo